### DISCUSSION TITLE
Python packaging references

### DISCUSSION THREAD

**Discussion by mdpiper** [csdms/project#23] [General]
The best way to share Python code with others is to _package_ it. Here are two references for Python packaging.

1. The [Python Packaging Authority](https://www.pypa.io/) is the standard. They have a [user guide](https://packaging.python.org/en/latest/) with a helpful [tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/). The [tutorial files](https://github.com/pypa/sampleproject), hosted on GitHub, show the directory structure of a Python package.
1. The [pyOpenSci Python Package Guide](https://www.pyopensci.org/python-package-guide/). This guide is created by Leah Wasser, a very experienced developer and educator, and a friend of CSDMS.

Here are two real-world examples: [one from Eric](https://github.com/csdms/bmi-wavewatch3), and [one from me](https://github.com/csdms/bmi-topography). Eric tends to do the right thing. I tend to do the least work possible.

If you'd like help packaging a Python project, please visit us at [CSDMS Office Hours](https://csdms.colorado.edu/wiki/OfficeHours)!

**Comment by samharrison7** [comment]
Do you guys ever use [Poetry](https://python-poetry.org/)? For pure Python packages it's pretty neat and does away with the need for a separate `setup.cfg` file - everything goes in `pyproject.toml` ([here's an example](https://github.com/microplastics-cluster/fragment-mnp/blob/develop/pyproject.toml)). It's scarily easy to build packages with it!

**Reply by mdpiper** [reply]
We've stuck with *setuptools* so far, though I'd have to defer to @mcflugen for future plans.

Currently, the only reason we have `setup.cfg` files is for *flake8*--it doesn't support configuration through `pyproject.toml`. We're working away from even this by using `.flake8`, or just setting up *pre-commit* rules on a repo.

**Reply by BSchilperoort** [reply]
> does away with the need for a separate `setup.cfg` file - everything goes in `pyproject.toml`

this is now standard in Python, across setuptools/hatchling/poetry/etc.

We often use [hatch](https://hatch.pypa.io/latest/)/hatchling as our build system. It tries to use common standards as much as possible to keep it easy to switch to different systems, allows for defining scripts and test matrices.

**Comment by hrajagers** [comment]
Hi Sam,

Yes, I'm using Poetry as well for two Python-based rapid assessment tools ([D-FAST Morphological Impact](https://github.com/Deltares/D-FAST_Morphological_Impact) and [D-FAST Bank Erosion](https://github.com/Deltares/D-FAST_Bank_Erosion)) to make sure that all developers (and build/test agents) are using the same environment. Since I'm compiling the tools before distribution using [Nuitka](https://nuitka.net/), I don't care too much about end users, but for advanced users (semi-developers) that want to run the tools inside Python it's equally useful.

**Reply by samharrison7** [reply]
I've never come across Nuitka before, that looks pretty useful!

**Comment by mcflugen** [comment]
@samharrison7 I tried poetry a while ago but decided it was a solution to a problem I didn't have. My current workflow with pip seems to work just fine.

Then when I discovered that *poetry* uses a random number generator to decide if it's going to succeed when running in CI (it's set to fail 5% of the time), I bailed. Maybe that's not the case anymore, I don't know, but it was (is) an odd decision by their developers.

I've also heard that *poetry* is slow when run in CI environments but I can neither confirm nor deny that claim.

Another option, which people these days seem to like, is *uv*. It's developed by *Astral* (the folks behind *ruff*), who I am not a fan of, so I'll probably continue to stick with my current workflow.

**Reply by samharrison7** [reply]
Thanks for the reflection. I've never used Poetry in CI so haven't come across these issues. `uv` looks interesting, they like to boast about their speed!

**Reply by hrajagers** [reply]
We are using `poetry` in a CI pipeline running on TeamCity and I've not encountered such random behaviour. I can't comment on whether it's the fastest or best solution; it's been good (enough?) for us.
